promql/parse.go

```diff
@@ -0,0 +1,864 @@
+	token := p.next()
+	if token.typ != expected {
+		p.unexpected(token, context)
+	}
+	return token
+}
+
+// expectOneOf consumes the next token and guarantees it has one of the required types.
+func (p *parser) expectOneOf(expected1, expected2 itemType, context string) item {
+	token := p.next()
+	if token.typ != expected1 && token.typ != expected2 {
+		p.unexpected(token, context)
+	}
+	return token
+}
+
+// unexpected complains about the token and terminates processing.
+func (p *parser) unexpected(token item, context string) {
+	p.errorf("unexpected %s in %s", token, context)
+}
+
+// recover is the handler that turns panics into returns from the top level of Parse.
+func (p *parser) recover(errp *error) {
+	e := recover()
+	if e != nil {
+		if _, ok := e.(runtime.Error); ok {
+			panic(e)
+		}
+		*errp = e.(error)
+	}
+	return
+}
+
+// stmt parses any statement.
+//
+// 		alertStatement | recordStatement
+//
+func (p *parser) stmt() Statement {
+	switch tok := p.peek(); tok.typ {
+	case itemAlert:
+		return p.alertStmt()
+	case itemIdentifier, itemMetricIdentifier:
+		return p.recordStmt()
+	}
+	p.errorf("no valid statement detected")
+	return nil
+}
+
+// alertStmt parses an alert rule.
+//
+//		ALERT name IF expr [FOR duration] WITH label_set
+//			SUMMARY "summary"
+//			DESCRIPTION "description"
+//
+func (p *parser) alertStmt() *AlertStmt {
+	const ctx = "alert statement"
+
+	p.expect(itemAlert, ctx)
+	name := p.expect(itemIdentifier, ctx)
+	// Alerts require a vector typed expression.
+	p.expect(itemIf, ctx)
+	expr := p.expr()
+
+	// Optional for clause.
+	var duration time.Duration
+	var err error
+
+	if p.peek().typ == itemFor {
+		p.next()
+		dur := p.expect(itemDuration, ctx)
+		duration, err = parseDuration(dur.val)
+		if err != nil {
+			p.error(err)
+		}
+	}
+
+	p.expect(itemWith, ctx)
+	lset := p.labelSet()
+
+	p.expect(itemSummary, ctx)
+	sum, err := strconv.Unquote(p.expect(itemString, ctx).val)
+	if err != nil {
+		p.error(err)
+	}
+
+	p.expect(itemDescription, ctx)
+	desc, err := strconv.Unquote(p.expect(itemString, ctx).val)
+	if err != nil {
+		p.error(err)
+	}
+
+	return &AlertStmt{
+		Name:        name.val,
+		Expr:        expr,
+		Duration:    duration,
+		Labels:      lset,
+		Summary:     sum,
+		Description: desc,
+	}
+}
+
+// recordStmt parses a recording rule.
+func (p *parser) recordStmt() *RecordStmt {
+	const ctx = "record statement"
+
+	name := p.expectOneOf(itemIdentifier, itemMetricIdentifier, ctx).val
+
+	var lset clientmodel.LabelSet
+	if p.peek().typ == itemLeftBrace {
+		lset = p.labelSet()
+	}
+
+	p.expect(itemAssign, ctx)
+	expr := p.expr()
+
+	return &RecordStmt{
+		Name:   name,
+		Labels: lset,
+		Expr:   expr,
+	}
+}
+
+// expr parses any expression.
+func (p *parser) expr() Expr {
+	return p.binaryExpr()
+}
+
+// binaryExpr parses a binary expression and changes precedence of sequential binary
+// expressions based on the operator's precedence.
+func (p *parser) binaryExpr() Expr {
```

Comment:
This name is weird, because it also parses non-binary simple expressions... right now `expr()` is just wrapping `binaryExpr()`. Maybe the wrapper should just go away and the whole thing be called `expr()`?